### PR TITLE
Closes #1879: remove temporary files

### LINKED ISSUE #1879: Remove some temporary files after they are not needed
We have some cases when we create temporary file but never remove them after the work is done.
Such files are in form:
```

tempSubsetFile2465843405109863067.tmp
tempLocalFile8286233180649395020.tmp
tempFileToRescale4901057174882206113.tmp
firstpass-2430429116362840534.tab
firstpass-2578408473252052700.csv
data-5170417214209956484.tab
d036468f-8051-406d-82bf-1928218066b82512574650660504461305af31c-5cd9-4e12-9a85-4071b9ec6e7d
tempTabfile.1257543624808568799.tab
dataverseDataFrame_867704.RData

```
We need to find code responsible for creating those and remove them after some processing is done.

### PR DESCRIPTION
**Pull request by dbojanek**

### CODE REVIEW

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/sav/SAVFileReader.java [hunk 1]
```diff
@@ -317,6 +318,13 @@ public SAVFileReader(TabularDataFileReaderSpi originator) {
         super(originator);
     }
 
+    @PreDestroy
+    private void cleanUp() {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/sav/SAVFileReader.java]: I don't think that this has any effect? `SavFileReader` is created using `new` so it is not a part of any context (either ejb nor cdi).
```
        } else if (mimeType.equals(ApplicationMimeType.SPSS_SAV.getMimeValue())) {
            ingestPlugin = new SAVFileReader(new SAVFileReaderSpi());
```

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java [hunk 2]
```diff
@@ -191,7 +190,11 @@ private boolean generatePDFThumbnail(StorageIO<DataFile> storageIO, int size, lo
             return false;
         } finally {
             tempThumbnailFile.delete();
-            tempPdfFile.ifPresent(file -> file.delete());
+            tempPdfFile.ifPresent(File::delete);
+            if (storageIO.isRemoteFile()) {
+                sourcePdfFile.ifPresent(File::delete);
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataaccess/ImageThumbConverter.java]: Didn't we already removed that file using previous solution? Notice that now `tempPdfFile == sourcePdfFile` if we have remote storage and if we have file storage tempPdfFile is empty and sourcePdfFile is not.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/DDIExportServiceBean.java [hunk 3]
```diff
@@ -456,19 +456,24 @@ private boolean checkIsWithoutFrequencies(List<DataVariable> vars) {
...
 
-            IngestServiceBean.produceFrequencies(tabFile, vars);
+            IngestServiceBean.produceFrequencies(tabFile.get(), vars);
 
         } catch (Exception ex) {
             logger.warning(ex.getMessage());
             return;
         } finally {
             tmpFile.ifPresent(file -> file.delete());
+            if (storageIO != null && storageIO.isRemoteFile()) {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/DDIExportServiceBean.java]: Similar case as some comment above. Did we have problems with deleting temporary file before this change?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/csv/CSVFileReader.java [hunk 4]
```diff
@@ -447,19 +447,18 @@ public int readFile(BufferedReader csvReader, DataTable dataTable, PrintWriter f
...
             }
+            long linecount = parser.getRecordNumber();
+            if (dataTable.getCaseQuantity().intValue() != linecount) {
+                List<String> args = Arrays.asList("" + dataTable.getCaseQuantity().intValue(),
+                        "" + linecount);
+                throw new IngestException(IngestError.CSV_LINE_MISMATCH, args);
+            }
+            return (int) linecount;
+        } finally {
+            firstPassTempFile.delete();
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/csv/CSVFileReader.java]: I don't think that will solve the issue completely. We are removing file only if we get to so-called second-pass, file is created before first pass. Unfortunately I can't think of "pretty" solution that would not be a complete rewrite of the class. 
Maybe we could move the creation of firstPassTempFile to the method above and pass it to this method.

```
File firstPassTempFile = File.createTempFile("firstpass-", ".csv");
try {
   int lineCount = readFile(localBufferedReader, dataTable, tabFileWriter, firstPassTempFile);
finally {
  firstPassTempFile.delete();
}
```

I know that the solution is not pretty but I can't think of a better way right now.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/xlsx/XLSXFileReader.java [hunk 5]
```diff
@@ -113,124 +113,126 @@ public TabularDataIngest read(BufferedInputStream stream, File dataFile) throws
...
-        }
-        dbglog.fine("Variable names: " + varNames);
+            dbglog.fine("Variable names: " + varNames);
 
 
-        return ingesteddata;
-
+            return ingesteddata;
+        } finally {
+            finalWriter.close();
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/xlsx/XLSXFileReader.java]: finalWriter could be easily included in try with resources statement

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/xlsx/XLSXFileReader.java [hunk 6]
```diff
@@ -113,124 +113,126 @@ public TabularDataIngest read(BufferedInputStream stream, File dataFile) throws
...
+            firstPassTempFile.delete();
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/tabulardata/impl/plugins/xlsx/XLSXFileReader.java]: Same issue as in CsvFileReader.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java [hunk 7]
```diff
@@ -917,16 +919,19 @@ public boolean ingestAsTabular(Long datafile_id) {
...
+                                tabDataIngest,
+                                originalFileSize,
+                                originalFileName,
+                                originalContentType);
 
                         dataFile = fileService.save(dataFile);
                     }
+                } finally {
+                    // delete the temp tab-file:
+                    tabFile.delete();
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:933]: A little better. But still quite possible for leaving some leftovers. Before the creation of this file and this delete statement there are quite few try catch blocks with return statements. Maybe lets leave it for now. As I think that when resolving issue with failure on bigger tabular files this method could be changed a little and deletion of file maybe will be a little easier.

**Review by madryk — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/DDIExportServiceBean.java (hunk 3 above)
Comment (dbojanek) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/DDIExportServiceBean.java]: I wasn't sure. I've tried to follow usage of `StorageIOUtils.obtainAsLocalFile`.

**Review by madryk — APPROVED** [review]

**Pull request merged**